Border panel: Update to display multiple palette origins

stacimc opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

Fixes #36641 
~~Alternative to #33743~~

## Description
<!-- Please describe what you have changed or added -->
The color panel is able to show three different palettes: the theme, the default, and the user one. This updates the border panel to provide the same options for border color.

* Extracts color and gradient settings retrieval to custom hook
* Update border panel to use palettes from multiple origins

~~@aaronrobertshaw implemented this fix in #33743 on top of work to switch the Border panel to use the ToolsPanel. That PR is currently blocked on a [discussion about how to handle defaults](https://github.com/WordPress/gutenberg/pull/33743#issuecomment-972637416) and base styles. The purpose of this PR is to have an alternative to fix the multiple palette issue in 5.9 just in case #33743 is not merged.~~

This fix was pulled out of #33743, which switches the Border panel to use the ToolsPanel. It is pulled out into its own PR for ease of getting it into 5.9, since it fixes #36641.

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

Color and gradient settings retrieval are extracted without change into a hook, but we should still retest the Colors panel using instructions from #35970 and #36492.

Test border color from all origins and verify that the behavior in the border panel matches the Colors panel.

## Screenshots <!-- if applicable -->
<img width="290" alt="Screen Shot 2021-11-22 at 12 45 37 PM" src="https://user-images.githubusercontent.com/63313398/142932798-172814ef-17bc-48e0-89e1-eb2424246c28.png">

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

aaronrobertshaw committed:
Extract color and gradient settings retrieval to custom hook

aaronrobertshaw committed:
Update border color to display multiple color palette origins

stacimc committed:
Merge in changes from trunk

aaronrobertshaw commented on packages/block-editor/src/hooks/border-color.js:
```diff
@@ -83,11 +89,10 @@ export function BorderColorEdit( props ) {
-			disableCustomColors={ disableCustomColors }
-			disableCustomGradients={ disableCustomGradients }
 			onColorChange={ onChangeColor }
+			clearable={ false }
```
I believe this `clearable` prop was only added to https://github.com/WordPress/gutenberg/pull/33743 because the control could be reset via the `ToolsPanel` menu. Given this PR is for before that lands perhaps it should be omitted here. 

It's only a small thing. I don't feel strongly either way.

andrewserong replied:
I don't feel strongly about it either. For consistency with the color panel, it'd make sense to omit it? In manual testing, you can select a theme or default color and then deselect it by clicking the circle to effectively clear the color setting, so I don't think it's a big issue if we leave it in, though.

aaronrobertshaw reviewed (commented):
Thanks for putting this together @stacimc 👍 

We definitely need it to make sure the border-colour control is brought up to date for 5.9. 

> Alternative to #33743

For others coming to this PR, it is more "related to" or a subset, rather than a full alternative to #33743. 

I've given this a quick test and it looks to be working for me. I did leave one tiny comment though that I'll leave up to you.

It would be good to get a second set of eyes on this to see if there is a better approach. Given my input in creating these changes a final tick of approval should probably be provided by someone else.

ramonjd commented:
I've been testing this and have noticed no real problems in the border panel color palette controls. They behave and look like the color panel controls.

<img width="289" alt="Screen Shot 2021-11-25 at 10 23 57 am" src="https://user-images.githubusercontent.com/6458278/143324935-e7094f58-401d-48e7-98ba-1d744caa6851.png">


> but we should still retest the Colors panel using instructions from #35970 and #36492.

Are there any preferred testing steps? Those PRs are very light on testing details, the latter contains none at all.

Also, I tried testing with a custom color palette in the Site Editor – 

![Nov-25-2021 10-29-19](https://user-images.githubusercontent.com/6458278/143325338-ceb18276-2475-49dc-8f0d-d8d0f8c7d0bc.gif)

– but it never saves. This is also happening on trunk so... 🤷

aaronrobertshaw commented:
> Also, I tried testing with a custom color palette in the Site Editor but it never saves. This is also happening on trunk so.

I've tested this as well but it is working for me on trunk. It looks like it was fixed yesterday when https://github.com/WordPress/gutenberg/pull/36817 was merged.

Prior to rebasing this PR I could replicate the issue, after rebasing, it is resolved. 

| Before Rebase | After Rebase |
|---|---|
| ![ColorPalettes-Before](https://user-images.githubusercontent.com/60436221/143387263-53e31138-a989-48f2-af8f-a2a2d9747e1d.gif) | ![ColorPalettes-After](https://user-images.githubusercontent.com/60436221/143387279-4a3348e0-4dc6-49d8-8ccb-7dfdf356496b.gif) |

andrewserong reviewed (approved):
This is testing nicely for me, with default and theme colors. It also still works well in a `theme.json` that doesn't provide palette colors and also switches off the default palette by setting `settings.color.defaultPalette` to `false` — here it's still possible to set a custom border color:

![image](https://user-images.githubusercontent.com/14988353/143793923-89f594a1-b521-4409-9d1a-d199f7024c60.png)

LGTM, and the regular color palettes still appear to be behaving correctly, and I like the refactor of adding in the `useMultipleOriginColorsAndGradients` hook — looks like it was safely moved over without changing anything 👍

oandregal commented:
Does this need to be backported to WordPress 5.9?

stacimc commented:
> Does this need to be backported to WordPress 5.9?

@oandregal Just added it to the Must-Haves project; I moved it directly to `Done` since it has been merged, let me know if that is incorrect. I updated the PR description as well for posterity: this work was pulled out of #33743 so that it would be easier to get a fix for #36641 into 5.9.

aaronrobertshaw commented:
> Does this need to be backported to WordPress 5.9?

To fix https://github.com/WordPress/gutenberg/issues/36641 for 5.9 so all the color controls display the multiple origin palettes I believe this should be backported. I've added the `Backport to WP Beta/RC` label.
